File: research/archive/20260909-file-tools-screen/file_tools.py
"""Independent file-interface activation and observed-mutation evidence."""
import copy
import json
from pathlib import Path

ALLOWED = {'run', 'job_start', 'job_output', 'job_status', 'job_wait', 'job_stop', 'update_plan'}


def transform(prompt):
    mapping = json.loads(Path(__file__).with_name('file_tools_prompt.json').read_text())
    for old, new in mapping.items():
        if prompt.count(old) != 1:
            raise ValueError('file prompt seam drift')
        prompt = prompt.replace(old, new)
    return prompt


def normalize(trajectory):
    """Keep raw evidence intact; insert observed mutations after command completion.

    A command which both edits and tests is deliberately not post-edit verification:
    the screen contract requires a separate final verification command in both arms.
    """
    result = copy.deepcopy(trajectory)
    events = []
    for event in result.get('events', []):
        events.append(event)
        if event.get('file_changes'):
            events.append({'name': 'mutation', 'phase': 'result',
                           'args': {'paths': event['file_changes']},
                           'result': {'is_error': False, 'content': 'Observed filesystem change'},
                           'file_boundary': event.get('file_boundary')})
    result['events'] = events
    return result


def audit(directory, mode, requests, trajectory):
    def read(name):
        return json.loads((directory / name).read_text())
    frozen = read('file-tools.json')
    if mode not in {'native', 'shell'} or frozen['mode'] != mode:
        raise ValueError('file tools mode mismatch')
    baseline = frozen['tools']
    if not baseline or any(t['type'] != 'function' or t['name'].startswith('mcp__') for t in baseline):
        raise ValueError('external or missing baseline tools')
    expected_tools = baseline if mode == 'native' else [t for t in baseline if t['name'] in ALLOWED]
    if not ALLOWED.issubset({t['name'] for t in expected_tools}):
        raise ValueError('missing preserved tools')
    expected_prompt = frozen['baseline'] if mode == 'native' else transform(frozen['baseline'])
    if frozen['effective'] != expected_prompt:
        raise ValueError('file prompt mismatch')
    for path in requests:
        request = json.loads(path.read_text())
        payload = read('provider-' + path.name)
        if request['system_prompt'] != expected_prompt or payload['instructions'] != expected_prompt or payload['tools'] != expected_tools:
            raise ValueError('file interface activation mismatch')
    previous = read('file-boundary-0000.json')
    completed = [e for e in trajectory.get('events', []) if e.get('result')]
    for index, event in enumerate(completed, 1):
        current = read(f'file-boundary-{index:04d}.json')
        changed = sorted(p for p in previous.keys() | current.keys() if previous.get(p) != current.get(p))
        if event.get('file_boundary') != index or sorted(event.get('file_changes', [])) != changed:
            raise ValueError('filesystem boundary evidence mismatch')
        if mode == 'shell' and event['name'] not in ALLOWED:
            raise ValueError('removed tool executed')
        previous = current

File: research/archive/20260909-file-tools-screen/file_tools.lua
-- Eval-only file-tool ablation and tool-neutral filesystem evidence.
local json = require('agent.util.json')
local uv = require('luv')
local registry = require('agent.tool_registry')
local M = {}
local allowed = {run=true, job_start=true, job_output=true, job_status=true,
    job_wait=true, job_stop=true, update_plan=true}
local function read(path)
    local f = assert(io.open(path, 'rb')); local s = f:read('*a'); f:close(); return s
end
local function save(path, value)
    local f = assert(io.open(path, 'w')); f:write(json.encode(value), '\n'); f:close()
end
function M.apply(mode, session, root, directory)
    assert(mode == 'native' or mode == 'shell', 'unsupported file tools')
    assert(session.tool_scope == 'local_only', 'file ablation requires local_only')
    local baseline = session:get_system_prompt()
    local effective = baseline
    local tools = registry.native_tools()
    for _, t in ipairs(tools) do assert(not t.name:match('^mcp__'), 'external tools not allowed') end
    if mode == 'shell' then
        local mapping = json.decode(read(root .. '/evals/file_tools_prompt.json'))
        for old, replacement in pairs(mapping) do
            local a, b = assert(effective:find(old, 1, true), 'missing prompt seam'), nil
            b = a + #old - 1
            assert(not effective:find(old, b + 1, true), 'duplicate prompt seam')
            effective = effective:sub(1,a-1) .. replacement .. effective:sub(b+1)
        end
        local original_tools, original_execute = registry.native_tools, registry.execute
        registry.native_tools = function()
            local selected = {}
            for _, t in ipairs(original_tools()) do
                if allowed[t.name] then selected[#selected+1] = t end
            end
            return selected
        end
        registry.execute = function(name, args, context)
            assert(allowed[name], 'file ablation blocked tool: ' .. tostring(name))
            return original_execute(name, args, context)
        end
    end
    session.system_prompt = effective
    save(directory .. '/file-tools.json', {mode=mode, baseline=baseline, effective=effective, tools=tools})
    local function snapshot()
        local files = {}
        local function scan(dir, prefix)
            local handle = assert(uv.fs_scandir(dir))
            while true do
                local name, kind = uv.fs_scandir_next(handle)
                if not name then break end
                if name ~= '__pycache__' and name ~= '.pytest_cache' and name ~= '.git' then
                    local relative = prefix .. name
                    if kind == 'directory' then scan(dir .. '/' .. name, relative .. '/')
                    elseif kind == 'file' then files[relative] = read(dir .. '/' .. name) end
                end
            end
        end
        scan(session.cwd, '')
        return files
    end
    local previous, index = snapshot(), 0
    save(directory .. '/file-boundary-0000.json', previous)
    return function(event)
        if not event.result then return end
        index = index + 1
        local current, changed = snapshot(), {}
        for path, body in pairs(previous) do if current[path] ~= body then changed[#changed+1] = path end end
        for path in pairs(current) do if previous[path] == nil then changed[#changed+1] = path end end
        table.sort(changed)
        event.file_changes = changed
        event.file_boundary = index
        save(directory .. string.format('/file-boundary-%04d.json', index), current)
        previous = current
    end
end
return M
